### PR TITLE
Backport lib/utils/prompt improvements to [v9]

### LINKED ISSUE #13021: Per-session MFA eats 1-2 keypresses (depending on devices)
### Expected behavior
Once the remote prompt appears, all input should get sent to the remote server.

### Current behavior
With "Per-session MFA" enabled, depending on what device(s) you have added, `tsh ssh` logins may eat 1-2 keypresses.

Scenarios:

1.  OTP and WebAuthn added
    1. OTP used = 1 keypress
    1. WebAuthn used = 2 keypresses
1. OTP only added = 1 keypress
1. WebAuthn only added = 0 keypress

It looks a bit similar to #11709.

#### Teleport version

9.3.0 (also tested with 9.2.4)

### Recreation steps

1. Enable "Per-session MFA"
2. Add an OTP device (and optionally a 2nd WebAuthn one)
3. Try to login in a device `tsh ssh`

### Debug logs

_Let me know if you would like them._

I'll be happy to help further, thanks!

### PR DESCRIPTION
**Pull request by codingllama**
`prompt.ContextReader`, along with various parts of `lib/utils/prompt`, where re-written in master so we can alleviate input swallowing issues.

For various reasons I didn't backport all changes to v9, but the less-eager input-swallowing loop is now what stands between us and issue #13021.

This isn't a backport of a specific PR, but instead a port of the entire `lib/utils/prompt` package, as the PRs that touch the package unfortunately do more than we want to backport. The interfaces are still compatible and I did test various `tsh login` and `tsh ssh` scenarios.

I've thrown in #13382 for good measure, as well.

Closes #13021.

### CODE REVIEW

**Comment by codingllama** [comment]
The initial commit is equivalent to:

```shell
rm lib/utils/prompt/*
git checkout master -- lib/utils/prompt/
```

`tsh` scenarios, including \<ctrl+c\> in pretty much every step. Showing so we can see the prompts.

```shell
$ ./tsh login --proxy=zarquon --user=llama
> Enter password for Teleport user llama:
> Tap any security key or enter a code from a OTP device:
> (ok)

$ ./tsh ssh --proxy=zarquon --user=llama llama@zarquon
> Tap any security key
> (ok, no input swallowed)

$ ./tsh ssh --proxy=zarquon --user=llama --mfa-mode=otp llama@zarquon
> Enter an OTP code from a device: <otp>
> (ok, no input swallowed)
```

File: lib/utils/prompt/context_reader_test.go [hunk 1]
```diff
@@ -0,0 +1,273 @@
...
+		buf, err := cr.ReadContext(ctx)
+		require.NoError(t, err)
+		require.Equal(t, string(buf), "hello")
+	})
+
+	t.Run("reclaim abandoned read", func(t *testing.T) {
+		done := make(chan struct{})
+		cancelCtx, cancel := context.WithCancel(ctx)
+		go func() {
+			time.Sleep(1 * time.Millisecond) // give ReadContext time to block
```
Comment (espadolini) [inline lib/utils/prompt/context_reader_test.go:56]: I don't recall if this was brought up in the PRs in master, but isn't this dangerously flaky? Was there no alternative to a `time.Sleep`?

Comment (codingllama) [reply lib/utils/prompt/context_reader_test.go:56]: I think we talked about this before, the test won't flake if ReadContext fails to block (you can remove the sleep and do a `go test -count=1000` and it'll be fine).  Ideally I want to cancel to hit _after_ ReadContext is blocked, but it's alright if it doesn't.

**Review by espadolini — APPROVED** [review]

**Review by ibeckermayer — APPROVED** [review]

**Comment by codingllama** [comment]
Thanks for the quick reviews, folks.

**Pull request merged**